### PR TITLE
Feature: Integration of Audio Vibes Logic with Machine Learning Model

### PR DESCRIPTION
**Pull request by BoyanXu**
### Summary
This pull request introduces a comprehensive feature that enhances the `dashboard` application's ability to analyze the mood of music tracks by integrating audio vibe analysis with the actual loading of the predictive model. The feature utilizes audio features data from the [Spotify API](https://developer.spotify.com/documentation/web-api/reference/get-audio-features) and a machine learning model to predict and present a mood assessment.

### Changes
- **dashboard/views.py**: Implemented `check_vibe` function to aggregate vibes from audio features and lyrics for a multifaceted mood analysis.
- **utils.py**: Created `deduce_audio_vibe`, a function that processes Spotify's audio features, predicts the mood, and identifies the most common vibe.
- **dashboard/apps.py**: Modified to include model loading during app readiness, ensuring that the predictive model is available for vibe analysis. The model will be only loaded once, so with [a low memory footprint](https://github.com/gcivil-nyu-org/Wednesday-Fall2023-Team-1/pull/110#discussion_r1376868162).

### Implementation Details
- `check_vibe` in `dashboard/views.py` now works with both lyrics and audio features to deliver a detailed vibe analysis to the user.
- `deduce_audio_vibe` in `utils.py` is the utility that performs the mood prediction using the pre-trained model and the Spotify data.
- `DashboardConfig` in `dashboard/apps.py` is enhanced to load the predictive model from a binary file, using `joblib`, when the application is ready.

### Added Benefits
- **Improved Analysis Accuracy**: The combined lyrical and audio analysis offers a more accurate depiction of the track's mood.
- **Scalability**: With the model now being loaded at the app level, it is easier to manage and scale the predictive feature across different components of the application.
- **Performance Optimization**: Loading the model only once during app initialization reduces overhead and enhances response times for vibe predictions.

### Testing
Testing should be directed to ensure that:
- The model loads successfully without affecting startup times significantly.
- The `check_vibe` function accurately combines lyrics and audio features to determine the track's mood.
- The system maintains its performance benchmarks with the new integrated analysis logic.

### Audio Vibes Machine Learning Model

This pull request introduces the integration of audio vibes logic into the `dashboard/views.py` file, utilizing a machine learning model trained to classify moods from audio features. The model's logic is encapsulated in the `utils.py` as the function `deduce_audio_vibe`.

#### Model Training and Evaluation Documentation
The `Gradient_Boosting_Classifier.bin` model, which this feature relies on, was trained in a rigorous environment addressing class imbalance with [SMOTE](https://machinelearningmastery.com/smote-oversampling-for-imbalanced-classification/) and optimizing for the highest performance. Below are the highlights of the training and evaluation process:

- **Data Preprocessing:**
  - Addressed class imbalance in the dataset with SMOTE, equalizing the representation across eight mood classes.
  - Normalized audio feature 'length' and prepared the dataset for training.

- **Model Training:**
  - The `GradientBoostingClassifier` was selected from a suite of classifiers for its superior performance.
  - Trained on a balanced dataset to ensure unbiased mood predictions.

- **Model Performance:**
  - Achieved an overall accuracy of 0.849 on the test dataset.
  - The classification report showcases high precision and recall across all mood classes, with the 'Gloomy' mood achieving perfect precision.
  - The model demonstrates a balanced f1-score distribution, indicating robustness in both precision and recall metrics.

#### Changes in Application Configuration
The application configuration has been updated in `dashboard/apps.py` to load the trained model upon the readiness of the application.

#### Code Changes
- **dashboard/views.py**: Added `check_vibe` function to combine lyric and audio vibe analysis.
- **utils.py**: Implemented `deduce_audio_vibe` for mood prediction from audio features.
- **dashboard/apps.py**: Updated to load the Gradient Boosting Classifier model.

#### Class Distribution Details
- **Before SMOTE:** `Counter({3: 153, 1: 130, ... 7: 49})`
- **After SMOTE:** `Counter({0: 153, 4: 153, ... 1: 153})`

#### Model Accuracy and Classification Report
```
Accuracy for Gradient Boosting Classifier : 0.849
Classification Report for Gradient Boosting Classifier : 
              precision    recall  f1-score   support

       Happy       0.77      0.71      0.74        28
         Sad       0.83      0.87      0.85        23
   Energetic       0.83      0.67      0.74        36
        Calm       0.88      1.00      0.93        28
     Anxious       0.80      0.86      0.83        37
    Cheerful       0.76      0.90      0.82        31
      Gloomy       1.00      0.94      0.97        31
     Content       0.96      0.87      0.92        31

    accuracy                           0.85       245
   macro avg       0.85      0.85      0.85       245
weighted avg       0.85      0.85      0.85       245
```

### CODE REVIEW

File: classifier.ipynb
Comment (yp2353) [inline classifier.ipynb:1]: Is it necessary to import this?

File: data_moods.csv
Comment (yp2353) [inline data_moods.csv:1]: Is it necessary to import this?

File: eight_moods.csv
Comment (yp2353) [inline eight_moods.csv:1]: Is it necessary to import this?

Comment (BoyanXu) [reply eight_moods.csv:1]: Dropped in #136

File: data_moods.csv
Comment (BoyanXu) [reply data_moods.csv:1]: Dropped in #136

File: classifier.ipynb
Comment (BoyanXu) [reply classifier.ipynb:1]: Dropped in #136

**Pull request closed without merging**